### PR TITLE
Add TableBuilderOptions::level and relevant changes

### PR DESCRIPTION
**Pull request by rockeet**
For issue: https://github.com/facebook/rocksdb/issues/1333

I also modified other relevant code.

### CODE REVIEW

**Comment by facebook-github-bot-9** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

File: table/table_builder.h [hunk 1]
```diff
@@ -54,15 +54,16 @@ struct TableBuilderOptions {
       CompressionType _compression_type,
```
Comment (IslamAbdelRahman) [inline table/table_builder.h]: Do you mind making `TableBuilderOptions` constructor `explicit`

File: tools/sst_dump_tool.cc [hunk 2]
```diff
@@ -199,10 +199,12 @@ int SstFileReader::ShowAllCompressionSizes(size_t block_size) {
     if (CompressionTypeSupported(i.first)) {
       CompressionOptions compress_opt;
       std::string column_family_name;
+      int unknown_level = -1;
       TableBuilderOptions tb_opts(imoptions, ikc, &block_based_table_factories,
                                   i.first, compress_opt,
                                   nullptr /* compression_dict */,
-                                  false /* skip_filters */, column_family_name);
+                                  false /* skip_filters */, column_family_name,
+                                  unknown_level);
```
Comment (IslamAbdelRahman) [inline tools/sst_dump_tool.cc]: What do you think about just passing -1 directly (without using a variable) and adding a comment /* unknown level */

Something like this
```
false /* skip_filters */, column_family_name,
-1 /* unknown level */);
```

**Comment by IslamAbdelRahman** [comment]
LGTM, Thanks @rockeet, I have some minor comments.
Can you also add a simple unit test if possible ?

**Comment by siying** [comment]
Passing level to table builder is fine to me. I think what is needed more is whether the file is the bottonmost level to me. I assume you only want to apply strong compression in the bottommost level, which you can get from Compaction::bottommost_level().

File: table/table_builder.h (hunk 1 above)
Comment (rockeet) [inline table/table_builder.h]: 'explicit' keyword for constructor is useful only when parameter number can be 1.

File: tools/sst_dump_tool.cc (hunk 2 above)
Comment (rockeet) [inline tools/sst_dump_tool.cc]: Thanks for your review, self explainable variable should be preferred to a comment. thought they are equivalent, as explained in: 
http://www.refactoring.com/catalog/extractVariable.html

**Comment by rockeet** [comment]
@IslamAbdelRahman Thanks for your review again. A unit test is not need now, because this code change does not impact any existing code, ---- there is no any existing code in rocksdb really using this newly added 'level'. The 'level' is only used in our custom TableFactory. Or you could give me a unit test scenario in rocksdb ?

**Comment by rockeet** [comment]
@siying I used `sub_compact->compaction->output_level()` for `TableBuilderOptions::level` and didn't use `Compaction::bottommost_level()`, is this right? ref:[https://github.com/rockeet/rocksdb/commit/88518a284e9996fd7e2e70d0fd4ccfe885190eae?diff=unified#diff-6e99ed9f03dfac4fb36ae1a1e5a770bbR1140](url)

File: table/table_builder.h (hunk 1 above)
Comment (IslamAbdelRahman) [inline table/table_builder.h]: You are right ! please ignore my comment

**Comment by IslamAbdelRahman** [comment]
@rockeet, I believe what @siying meant was that the TableBuilder will now know the level number but what you really want to know is if this level is the bottommost level or not, and you can know what is the bottommost level by checking Compaction::bottommost_level()

**Comment by siying** [comment]
@rockeet what you are doing is correct. I was just saying that, you could use a better indication of whether you need to use your compression, which is whether the file is for the bottom-most level.

**Comment by facebook-github-bot-4** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Comment by IslamAbdelRahman** [comment]
Thanks @rockeet !

**Pull request merged**